java JNI code compiler warnings

There are a number of compiler warnings emitted when compiling the native methods for the MPI java bindings.  I used the following compiler options:

./configure --enable-mpi-java --enable-picky

Examples of the warnings:

mpi_MPI.c:81:1: warning: missing braces around initializer [-Wmissing-braces]
 opal_free_list_t ompi_java_buffers = {{0}};
 ^
mpi_MPI.c:81:1: warning: (near initialization for 'ompi_java_buffers.super.super') [-Wmissing-braces]
mpi_MPI.c: In function 'ompi_java_forgetDatatypeArray':
mpi_MPI.c:1024:13: warning: comparison of distinct pointer types lacks a cast
     if(jptr != cptr)
             ^
  CC       mpi_Prequest.lo
  CC       mpi_Status.lo
  CC       mpi_Win.lo
mpi_Win.c: In function 'Java_mpi_Win_compareAndSwap':
mpi_Win.c:449:63: warning: passing argument 4 of 'MPI_Compare_and_swap' makes pointer from integer without a cast
     int rc = MPI_Compare_and_swap(orgPtr, compPtr, resultPtr, dataType, targetRank, targetDisp, (MPI_Win)win);
                                                               ^
In file included from mpi_Win.c:30:0:
../../../../ompi/include/mpi.h:1339:20: note: expected 'MPI_Datatype' but argument is of type 'jlong'
 OMPI_DECLSPEC  int MPI_Compare_and_swap(const void *origin_addr, const void *compare_addr,

--- a/ompi/mpi/java/c/mpi_MPI.c
+++ b/ompi/mpi/java/c/mpi_MPI.c
@@ -78,7 +78,7 @@
 
 ompi_java_globals_t ompi_java = {0};
 int ompi_mpi_java_eager = 65536;
-opal_free_list_t ompi_java_buffers = {{0}};
+opal_free_list_t ompi_java_buffers = {{{0}}};
 static void *libmpi = NULL;
 
 static void bufferConstructor(ompi_java_buffer_t *item)
@@ -1021,7 +1021,7 @@ void ompi_java_getDatatypeArray(JNIEnv *env, jlongArray array,
 void ompi_java_forgetDatatypeArray(JNIEnv *env, jlongArray array,
                               jlong *jptr, MPI_Datatype *cptr)
 {
-    if(jptr != cptr)
+    if((long)jptr != (long)cptr)
         free(cptr);
 
     (*env)->ReleaseLongArrayElements(env, array, jptr, JNI_ABORT);

--- a/ompi/mpi/java/c/mpi_Win.c
+++ b/ompi/mpi/java/c/mpi_Win.c
@@ -446,7 +446,8 @@ JNIEXPORT void JNICALL Java_mpi_Win_compareAndSwap (JNIEnv *env, jobject jthis, 
     void *compPtr = (*env)->GetDirectBufferAddress(env, compareAddr);
     void *resultPtr = (*env)->GetDirectBufferAddress(env, resultAddr);
 
-    int rc = MPI_Compare_and_swap(orgPtr, compPtr, resultPtr, dataType, targetRank, targetDisp, (MPI_Win)win);
+    int rc = MPI_Compare_and_swap(orgPtr, compPtr, resultPtr, (MPI_Datatype)dataType, 
+	targetRank, targetDisp, (MPI_Win)win);
     ompi_java_exceptionCheck(env, rc);
 }
 
@@ -457,7 +458,8 @@ JNIEXPORT void JNICALL Java_mpi_Win_fetchAndOp(JNIEnv *env, jobject jthis, jlong
     void *resultPtr = (*env)->GetDirectBufferAddress(env, resultAddr);
     MPI_Op op = ompi_java_op_getHandle(env, jOp, hOp, baseType);
 
-    int rc = MPI_Fetch_and_op(orgPtr, resultPtr, dataType, targetRank, targetDisp, op, (MPI_Win)win);
+    int rc = MPI_Fetch_and_op(orgPtr, resultPtr, (MPI_Datatype)dataType, targetRank, 
+	targetDisp, op, (MPI_Win)win);
     ompi_java_exceptionCheck(env, rc);
 }
 
@@ -477,7 +479,7 @@ JNIEXPORT void JNICALL Java_mpi_Win_setName(
         JNIEnv *env, jobject jthis, jlong handle, jstring jname)
 {
     const char *name = (*env)->GetStringUTFChars(env, jname, NULL);
-    int rc = MPI_Win_set_name((MPI_Comm)handle, (char*)name);
+    int rc = MPI_Win_set_name((MPI_Win)handle, (char*)name);
     ompi_java_exceptionCheck(env, rc);
     (*env)->ReleaseStringUTFChars(env, jname, name);
 }
@@ -487,7 +489,7 @@ JNIEXPORT jstring JNICALL Java_mpi_Win_getName(
 {
     char name[MPI_MAX_OBJECT_NAME];
     int len;
-    int rc = MPI_Win_get_name((MPI_Comm)handle, name, &len);
+    int rc = MPI_Win_get_name((MPI_Win)handle, name, &len);
 
     if(ompi_java_exceptionCheck(env, rc))
         return NULL;

--- a/oshmem/shmem/java/java/Addr.java
+++ b/oshmem/shmem/java/java/Addr.java
@@ -26,6 +26,7 @@ private Addr()
  * Allocates a block of memory in the symmetric heap of the calling PE.
  * <p>Java binding of {@code shmalloc}.
  * @param size Size of the requested memory block, in bytes.
+ * @throws ShMemException Allocation error.
  */
 public Addr(int size) throws ShMemException
 {
@@ -41,6 +42,7 @@ private native ByteBuffer malloc(int size);
  * <p>Java binding of {@code shmemalign}.
  * @param align Size of the alignment block, in bytes.
  * @param size  Size for the memory block, in bytes.
+ * @throws ShMemException Allocation error.
  */
 public Addr(int align, int size) throws ShMemException
 {
@@ -54,6 +56,7 @@ private native ByteBuffer memalign(int align, int size);
  * Expands or reduces the size of the block.
  * <p>Java binding of {@code shrealloc}.
  * @param size New size for the memory block, in bytes.
+ * @throws ShMemException Allocation error.
  */
 public void realloc(int size) throws ShMemException
 {

--- a/oshmem/shmem/java/java/PSync.java
+++ b/oshmem/shmem/java/java/PSync.java
@@ -10,6 +10,7 @@ private long handle;
 /**
  * Allocates a symmetric work array.
  * @param size Number of elements in the work array.
+ * @throws ShMemException Allocation error.
  */
 public PSync(int size) throws ShMemException
 {

